CDCgov/prime-simplereport
# use okta filter to fetch all groups

# BACKEND PULL REQUEST

## Related Issue
- #4161 
- Organizations with a large amount of facilities experience errors when adding/modifying users facility access because some Okta groups have not been loaded.
- Example - Inviting a user to a group that has not been loaded. Typically these are the very first groups on the list.
![image](https://user-images.githubusercontent.com/10108172/184696760-9c1d55e7-851c-4958-91d1-113609038edd.png)
- Example - Adding a user to a group that has not been loaded will throw a "Something went wrong error" instead of a popup. This will only happen occasionally on `main`.
![image](https://user-images.githubusercontent.com/10108172/184696195-a5f4d2bd-57b8-41ce-bd45-8b04fd4ec155.png)


## Changes Proposed

- Using an updated version of the Okta SDK, we can utilize the [search parameter](https://developer.okta.com/docs/reference/api/groups/#request-parameters-3) instead of the 'q' parameter which seems to limit the amount of results. 


## Additional Information

- Okta SDK was upgraded to latest (from 3.0.2 to 8.2.1)
- The ability to use a search filter & q was removed in a newer version of the SDK, so some queries had to be updated to keep only one of them. 


Other funkiness
- `q` parameter only provides 300 results and getNextPage is null
![image](https://user-images.githubusercontent.com/10108172/184718748-f037863b-fe12-4af0-a809-6496d10cd77b.png)
![image](https://user-images.githubusercontent.com/10108172/184718820-4e63e7da-10d0-4976-abf8-0e32790ece1e.png)
vs `search` results
![image](https://user-images.githubusercontent.com/10108172/184718691-16cf66f3-c01b-44a5-b3e2-f4889d9a026f.png)

- `search` parameter may skip new groups that we were recently added. This seems to mostly be a problem when creating a new organization
![image](https://user-images.githubusercontent.com/10108172/184718052-3990aa44-e9e5-4857-a1d2-526b6cbd1b99.png)
evaluating the expression a few seconds later
![image](https://user-images.githubusercontent.com/10108172/184718330-941d629f-a6ff-4996-a04e-2aa01931c080.png)
vs `q` results 
![image](https://user-images.githubusercontent.com/10108172/184718422-1add61fa-ce25-4787-af6c-a59908cca75a.png)

## Testing
- Attempt to add a new user to Boban's Org with quite a few facilities checked
- Attempt to update an existing user with a few facilities checked (an error will occur randomly on "main", should not occur at all on the new version).

- Regression testing has been completed for LiveOktaRepository
  - Created new organization
  - Added new facilities
  - Adding new user to one facilitity
  - Adding new user to all organizations (had to increase the max size of the array for graphql validation)
  - Updating existing user to multiple facilities
- Regression testing for LiveOktaAuthentication
  - Get user-status,
  - Set password
  - Set recovery question
  - Enroll email mfa
  - Enroll in Phone call
  - ❌ Enroll in Google authenticator - qr code doesn't show up, but signing up with the link in the console works!
![image](https://user-images.githubusercontent.com/10108172/185423354-2bcf9138-28dc-474d-a9a2-dd1907fd724a.png)
  - ❌ Activate Security Key - I don't have one to use 
  - Verified Passcode



## Checklist for Primary Reviewer

- [ ] GraphQL schema changes are backward compatible with older version of the front-end
